### PR TITLE
Adding shellcheck

### PR DESCRIPTION
**Pull request by dbast**
Fixes https://github.com/conda-forge/staged-recipes/issues/4723

Build scripts inspired by https://github.com/conda-forge/git-annex-feedstock

### CODE REVIEW

File: recipes/shellcheck/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - toolchain
+    - stack
+
+  run:
+    - gmp 6.1.*
```
Comment (jakirkham) [inline recipes/shellcheck/meta.yaml]: Would relax the pinning as we add this automatically. Also please add a `host` section with `gmp` as well.

File: recipes/shellcheck/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  url: https://github.com/koalaman/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - toolchain
```
Comment (jakirkham) [inline recipes/shellcheck/meta.yaml]: Please update to the use the new `compiler` syntax.

ref: https://conda-forge.org/docs/meta.html#compilers

File: recipes/shellcheck/build.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+set -o xtrace -o pipefail -o errexit
+
+BINARY_HOME=${PREFIX}/bin
+PACKAGE_HOME=${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}
+
+mkdir -p ${BINARY_HOME}
+mkdir -p ${PACKAGE_HOME}
+
+export LIBRARY_PATH=${LIBRARY_PATH}:${PREFIX}/lib
```
Comment (jakirkham) [inline recipes/shellcheck/build.sh]: Why is this needed?

**Comment by jakirkham** [comment]
Thanks for the submission, @dbast. Added a few comments above. Please let us know if you have any questions. 😄

**Comment by dbast** [comment]
@jakirkham Thanks for your input. Yes there are still a lot of details to be clarified. The recipe is based on the only other Haskel feedstock I found: https://github.com/conda-forge/git-annex-feedstock Will need to strip it down and simplify if possible. Thats why it is Work in Progress WIP ;)

File: recipes/shellcheck/build.sh (hunk 3 above)
Comment (dbast) [reply recipes/shellcheck/build.sh]: Good point. It from https://github.com/conda-forge/git-annex-feedstock Need to clarify this.

File: recipes/shellcheck/meta.yaml (hunk 1 above)
Comment (dbast) [reply recipes/shellcheck/meta.yaml]: Ok

Comment (dbast) [reply recipes/shellcheck/meta.yaml]: For the host section I am not sure. The dependency is not coming from the package itself. It is coming via the compiler (stack). That is like {{native}}gcc-libs, which is only added to run in many recipes, but not to host.

**Comment by dbast** [comment]
@jakirkham Reworked the build scripts, added windows and made the CIs run through :)

There are some lib warnings, "/lib64/libdl.so.2 ... not found in sysroot, is this binary repackaging? .. do you need to use install_name_tool/patchelf?" <- what to do about those, ignore them?

I didn't got it building completely static by adding "-optl-static" like in https://github.com/koalaman/shellcheck/blob/master/Dockerfile (requires that all libs are also static?)

Pinging also @notestaff @yarikoptic the maintainers of the only "stack" built conda-forge package as of now. Maybe you have some review input here. Thanks

File: recipes/shellcheck/bld.bat [hunk 4]
```diff
@@ -0,0 +1,28 @@
+set "BINARY_HOME=%PREFIX%\bin"
+set "PACKAGE_HOME=%PREFIX%\share\%PKG_NAME%-%PKG_VERSION%-%PKG_BUILDNUM%"
```
Comment (jakirkham) [inline recipes/shellcheck/bld.bat]: Is this where `stack` packages are being installed? Should we do something like `%PREFIX%\share\stack\...`?

File: recipes/shellcheck/bld.bat [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      --extra-lib-dirs "%PREFIX%\lib" ^
+      --stack-root "%STACK_ROOT%" ^
+      setup ^
+      || goto :error
+stack --local-bin-path "%PREFIX%\bin" ^
+      --extra-include-dirs "%PREFIX%\include" ^
+      --extra-lib-dirs "%PREFIX%\lib" ^
+      --stack-root "%STACK_ROOT%" ^
+      install --ghc-options ^
+        "-optl-pthread -optl-L%PREFIX%\lib -optl-Wl,-rpath,%PREFIX%\lib" ^
```
Comment (jakirkham) [inline recipes/shellcheck/bld.bat]: Hmm...what does `-rpath` do on Windows? 🤔

File: recipes/shellcheck/build.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+set -o xtrace -o pipefail -o errexit
+
+BINARY_HOME=${PREFIX}/bin
+PACKAGE_HOME=${PREFIX}/share/${PKG_NAME}-${PKG_VERSION}-${PKG_BUILDNUM}
```
Comment (jakirkham) [inline recipes/shellcheck/build.sh]: Same question above about where to place `stack` packages.

**Comment by jakirkham** [comment]
Thanks @dbast. Appreciate the effort here. This is looking really good.

Had a couple questions about organization of packages above.

> There are some lib warnings, "/lib64/libdl.so.2 ... not found in sysroot, is this binary repackaging?

The warnings are interesting as `libdl` is system provided and TBH I can't really see a situation where the system would not included it or (for that matter) why we would want to package it. If you could raise an issue over at [`conda-build`]( https://github.com/conda/conda-build/ ) and ping me on it, that would be really helpful.

> do you need to use install_name_tool/patchelf?" <- what to do about those, ignore them?

`conda-build` already installs these tools as needed and makes use of them when wrapping up the package. So we shouldn't need to use them directly. In fact the warnings above are issued during this process.

> I didn't got it building completely static by adding "-optl-static" like in https://github.com/koalaman/shellcheck/blob/master/Dockerfile (requires that all libs are also static?)

Didn't totally follow the question, but will try to respond to as best I can.

It would be nice to use dynamic linking wherever possible. That is one of the major benefits of a package manger like `conda` is keeping packages small with limited focus. That said, in Go, static linking is the norm, which means we have had to embrace it for Go packages. My knowledge of Haskell packaging is rather limited. So would be interested to hear what experts in that area have to say.

File: recipes/shellcheck/bld.bat (hunk 5 above)
Comment (dbast) [reply recipes/shellcheck/bld.bat]: I guess that does nothing and can be removed. Will try that.

File: recipes/shellcheck/bld.bat (hunk 4 above)
Comment (dbast) [reply recipes/shellcheck/bld.bat]: Hm, it is not a stack package. It is a Haskel package compiled via GHC... stack just creates the environment to build it. That's why I don't see `stack` as part of directory name. This path would theoretical allow to have several packages in different versions/build numbers in that path. That is anyway not used and at the end of the build file only the resulting binary is kept and the whole `PACKAGE_HOME` folder is deleted before conda could archive it.

**Comment by dbast** [comment]
@jakirkham Opened a ticket in conda-build as suggested.

Some answers as best as I can to your questions see above. The build files origin from the git-annex feedstock and some stuff in it like `PACKAGE_HOME` and potentially not required lib/include directories could be most probably simplified, but this way serve as blue print for more complex future Haskell packages.

I know also did some more optimizations to shrink the binary size, which are in line with https://github.com/koalaman/shellcheck/blob/master/Dockerfile ... but not all compiler options are available on win-64, osx.

Regarding static/non-static: I agree with you, dynamic linking wherever possible. Independently from  that, I found out the following regarding Haskell: With the right compiler options (split/gc-sections) + strip it seems to be possible to build reasonable small full static Haskell binaries like with GO or Rust (but only on Linux). Without those options the binaries are quite huge (25-45Mbyte) even though they are not static.

File: recipes/shellcheck/bld.bat (hunk 4 above)
Comment (jakirkham) [reply recipes/shellcheck/bld.bat]: It might be worth raising an issue over at the `stack` feedstock so that we can determine some best practices for compiling Haskell packages and wrap them up in a `compiler` that everyone can use easily. This is a long term project, but it would really help the Haskell ecosystem in conda-forge to flourish.

**Comment by jakirkham** [comment]
Thanks @dbast. This is in.

There are still some things I'm not totally sure about, but I'm also not an expert on building Haskell packages. Let's go ahead and get this out there. Work on a Haskell compiler package akin to what was done with Go recently should help cutdown on this boilerplate.

**Comment by dbast** [comment]
Thx. Will follow the Haskell compiler package updates in the future to reduce the boilerplate.
